dosbox-staging/dosbox-staging
# Use atomic operations and busy wait for Voodoo threads

# Description

- Fix an assertion failure when `voodoo_threads = 1`
- Removes the Semaphore class and replaces it with atomic operations.  You can call it busy waiting but lock-free programming sounds better.  There's no more locks in the Voodoo code after this 😄
- No more assigning chunks to specific threads.  There is now an atomic chunk index.  Whoever gets there first does the work.  This can help if the OS scheduler sends a thread out to lunch for a while.
- `voodoo_threads = auto` is now based on physical cores rather than logical cores.  I see a performance decrease when number of threads exceeds physical cores.  Since the threads are spinning all the time, hyperthreading hurts more than it helps.


## Related issues

Fixes #3947

# Release notes

Performance increase for 3dfx Voodoo emulation

# Manual testing

Benchmark numbers are about the same as I previously reported here:  https://github.com/dosbox-staging/dosbox-staging/pull/3935#issuecomment-2357320164

Tested with `taskset -c 0-7 dosbox` to simulate a quad core with hyperthreading by setting process affinity.  On my system, the hyperthread siblings are sequential so logical cores 0 and 1 share physical core ID 0 and so on.  This can vary based on CPU so check `/proc/cpuinfo` if you're on Linux.

If someone with an actual 2 or 4 core CPU with hyperthreading wants to test and see if more threads helps or hurts performance, that would be useful information.


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
